apiserver/billing/rbac.go

```diff
@@ -0,0 +1,91 @@
+package billing
+
+import (
+	"context"
+	"fmt"
+
+	rbacv1 "k8s.io/api/rbac/v1"
+	apimeta "k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apiserver/pkg/endpoints/filters"
+	"k8s.io/apiserver/pkg/registry/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type storageCreator interface {
+	rest.Storage
+	rest.Creater
+	rest.Scoper
+}
+
+// createRBACWrapper is a wrapper around the storage that creates a ClusterRole and ClusterRoleBinding for each BillingEntity on creation.
+type createRBACWrapper struct {
+	storageCreator
+	client client.Client
+}
+
+func (c *createRBACWrapper) Create(ctx context.Context, obj runtime.Object, createValidation rest.ValidateObjectFunc, opts *metav1.CreateOptions) (runtime.Object, error) {
+	attr, err := filters.GetAuthorizerAttributes(ctx)
+	if err != nil {
+		return nil, err
+	}
+	user := attr.GetUser()
+
+	createdObj, err := c.storageCreator.Create(ctx, obj, createValidation, opts)
+	if err != nil {
+		return createdObj, err
+	}
+
+	ac := apimeta.NewAccessor()
+
+	objName, err := ac.Name(createdObj)
+	if err != nil {
+		return createdObj, err
+	}
+
+	rolename := fmt.Sprintf("billingentities-%s-viewer", objName)
+
+	role := &rbacv1.ClusterRole{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: rolename,
+		},
+		Rules: []rbacv1.PolicyRule{
+			{
+				APIGroups:     []string{"rbac.appuio.io"},
+				Resources:     []string{"billingentities"},
+				Verbs:         []string{"get"},
+				ResourceNames: []string{objName},
+			},
+		},
+	}
+
+	rolebinding := &rbacv1.ClusterRoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: rolename,
+		},
+		Subjects: []rbacv1.Subject{
+			{
+				Kind:     "User",
+				APIGroup: "rbac.authorization.k8s.io",
+				Name:     user.GetName(),
+			},
+		},
+		RoleRef: rbacv1.RoleRef{
+			Kind:     "ClusterRole",
+			APIGroup: "rbac.authorization.k8s.io",
+			Name:     rolename,
+		},
+	}
+
+	err = c.client.Create(ctx, role, &client.CreateOptions{DryRun: opts.DryRun})
+	if err != nil {
+		return createdObj, err
+	}
+	err = c.client.Create(ctx, rolebinding, &client.CreateOptions{DryRun: opts.DryRun})
+	if err != nil {
+		return createdObj, err
+	}
+
+	return createdObj, nil
+}
```

Comment:
When creating the object in odoo succeeds but one of the later calls fails we end up in an inconsistent state where the object in odoo exists, but the necessary RBAC doesn't. That's not terrible as this should be fairly unlikely, but we need to be aware of that. 

There is a similar problem in the organization API server where I tried to mitigate the issue by attempting a rollback. A proper fix would be do this in a controller.

Reply:
It's note done in a controller since we're not sure if that works without watch and how quickly the RBAC rules are updated. But yes ideally it would have both this and a controller.

I'll implement a rollback tho.